mllib/src/main/scala/org/apache/spark/ml/feature/LSH.scala

```diff
@@ -138,13 +143,13 @@ private[ml] abstract class LSHModel[T <: LSHModel[T]]
       val hashDistCol = hashDistUDF(col($(outputCol)))
 
       // Compute threshold to get exact k elements.
-      // TODO: SPARK-18409: Use approxQuantile to get the threshold
-      val modelDatasetSortedByHash = modelDataset.sort(hashDistCol).limit(numNearestNeighbors)
-      val thresholdDataset = modelDatasetSortedByHash.select(max(hashDistCol))
-      val hashThreshold = thresholdDataset.take(1).head.getDouble(0)
+      val quantile = numNearestNeighbors.toDouble / modelDataset.count()
+      val modelDatasetWithDist = modelDataset.withColumn(distCol, hashDistUDF(col($(outputCol))))
+      val hashThreshold = modelDatasetWithDist.stat
+        .approxQuantile(distCol, Array(quantile), $(relativeError))
 
       // Filter the dataset where the hash value is less than the threshold.
-      modelDataset.filter(hashDistCol <= hashThreshold)
+      modelDatasetWithDist.filter(hashDistCol <= hashThreshold(0))
```

Comment:
since the threshold is approximate, we still need to put a limit here to return no more `numNearestNeighbors` items.

Reply:
I guess maybe we don't need to put a limit here?  There is a limit later in the code. 
```
    val modelSubsetWithDistCol = modelSubset.withColumn(distCol, keyDistUDF(col($(inputCol))))
    modelSubsetWithDistCol.sort(distCol).limit(numNearestNeighbors)
```

Reply:
Hm, I guess it's also possible we get too few nearest neighbors. This is probably especially likely as the quantile is small. It may be a good idea to request too many nearest neighbors, to make the likelihood of returning too few pretty small.

On that note, is it meaningful to expose relativeError to callers? they want a number of nearest neighbors, not more or less. This is a pretty internal implementation detail. How about simply setting some fixed value, plus oversampling, which should virtually always give enough results yet gets some efficiency gains? I don't know what that value is; might bear a little testing.

Reply:
@huaxingao ah, I saw it. yea, I think we don't need more limit. Probably we need oversampling to get more neighbors in case of what @srowen said.

Reply:
@srowen Thanks for your review. I will remove the exposure of relativeError. Will also do some tests to find out a good number to use for oversampling.